### DISCUSSION TITLE
issues getting data from big Atom

### DISCUSSION THREAD

**Discussion by lveillard** [pmndrs/jotai#2118] [Q&A]
So we have a big Atom and we get it using selectArom

```
export const fieldAtom = (fieldId: string) =>
  selectAtom(
    fieldMapAtom,
    (fieldMap) => {
      console.log(`field: ${fieldId} updated`);
      return fieldMap.get(fieldId) || undefined;
    },
    deepEqual
  );
```

Then we call this selctor in other atoms, and in components (with useMemo).

The dependencies are ok, they don't spread, however the selectAtom, even without being called, is recomputed on every fieldMapAtom change, which means that ```     console.log(`field: ${fieldId} updated`);``` gets called once per fieldID, and i guess it happens also for that deepEqual operation.

This adds some performance issues.

1) Is there a way to get a subset of the map in atoms and components without the reactive part? Without deepEqual requirement? Obviously i want the atoms and component depending on that field to get updated, but i don't want the intermediary compute to catch  and filter every change in the big map, I want it to jut subscribe to a part of the bigAtom
2) wouldn't it be better to generate this stores as individual atoms instead of a big Map with selector? If yes, how could it be done?

**Comment by dai-shi** [comment]
I have a question: Do you need `deepEqual`? Because if you need it, it means you mutate fieldMap value (violating the immutability contract), which isn't good for performance in some cases.

There are probably several approaches including your second point, but I think you want to make it sure first if your data is immutable or it can't be so for some reason.

**Reply by lveillard** [reply]
I think we do need deepEqual but i'm not sure. This is our use case:
 
 - FieldMap is a cache that gets filled with different graphQL queries 
 - Those fields hold an expression (excel like formula) and other rules. 
- Field's expressions might depend on other fields.
- If the expression changes we want every upward dependency to recompute. 
- A family atom computes the values of those expressions: so computeValueByFieldIdFamilyAtom might depend on one or multiple field's expressions 
- computeValueByFieldIdFamilyAtom can call other computeValueByFieldIdFamilyAtom for instance `isAdult <= Age <= Birthday` (that's why we made them family atoms)
- Fields are dynamically generated and can change on real time, so the fields might not only change the expression but also fields can be added or deleted

When the fieldMap is updated, we save it as a new map with a new reference in order to spread the changes. 

**Example**:
Imagine the fields:
- $isAdult  ( $Age > 18 )
- $Age (TODAY -$birthday)
- $Birthday  (and editable value)

If we change the birthday value, or the expression (in the fieldMap) of any of the other two fields, we need the upwards dependencies to recompute. Other fields don't need a recompute.

The problem here is that if we change a different field, the fieldMap reference is updated, all the selectAtoms are recomputed in order to filter the dependencies, deepEqual is used in all those selectAtoms and while only the right atoms in the value atom family are updated (as expected), all the deepEqual functions are fired.

**Reply by dai-shi** [reply]
> When the fieldMap is updated, we save it as a new map with a new reference in order to spread the changes.

If everything is updated immutably, I don't think you need deepEqual and you don't need to use selectAtom. selectAtom is a special util and it should call equality function every time.

**Reply by lveillard** [reply]
Hey! Nor sure if i got it 🤔

The value of each atom in the value atomfamily depends on one or N fields' expressions in the fieldAtom. 

Without the select atom, if a field in the fieldMap changes, all the atoms computing the values get recomputed, not only the ones impacted by the changed field.

So for instance when the name of the field "family name" changes the values depending on the field "age" get recomputed, while that property and that field are not a real dependency of family name's value atoms

**Comment by dai-shi** [comment]
Does this work?

```ts
export const fieldAtom = (fieldId: string) =>
  atom((get) => get(fieldMapMap).get(fieldId) || undefined);
```

**Reply by lveillard** [reply]
it does work but it spreads computes to non dependent fields, here is a working example:
https://codesandbox.io/s/distracted-hill-cp7xjl?file=/src/atoms.js

When clicking on the button, only birthday is changing, so only age should be recomputed

**Reply by lveillard** [reply]
Oh wait, I was adding all the values as dependencies. It seems that it works!  Lets leverage the example code to clarify two questions:
1) If those () => atom do work, what would be the purpose of the selectAtom instead of using just this?
2) Any way to avoid using the atomFamily for the values knowing that the field list is dynamic ) so we can't manually pre-define each atom)

**Reply by dai-shi** [reply]
> If those () => atom do work, what would be the purpose of the selectAtom instead of using just this?

selectAtom is only useful when custom `equalifyFn` is required. But using it is generally not recommended. It's an escape hatch.

> Any way to avoid using the atomFamily for the values knowing that the field list is dynamic ) so we can't manually pre-define each atom)

Not exactly sure what you mean, but an alternative to atomFamily is atoms-in-atom pattern: https://jotai.org/docs/guides/atoms-in-atom

**Reply by lveillard** [reply]
So while in the codesandbox the `(param) => atom(... `seem to work, for some reason when migrating some of the select atoms to that structure we actually suddenly get infinite rerenders. Im not sure why think will try to reproduce it in a codesandbox.


Regarding the atom-in-atom im not sure how we could achieve it as we don't know beforehand the number of atoms. 

The computed values work fine using family atoms because basically it creates one atom per field, and it adds new atoms om demand when new fields appear. 


So for instance if we have 500 fields and in a form we have 4 components nourishing 4 editable fields and wr have 2 computed fields that dépend on some of those values, the atom family will generate 6 atoms. 

If suddenly the form gets a new component with a new field for sone reason, the atomfamily generates a 7th atom. Those atoms coould not be generated on build time, they generated on runtime depending on some dynamic json files / database queries. 

Not sure how to create an atom-in-atom pattern that grows/shrinks on demand on runtime 🤔

**Reply by dai-shi** [reply]
The todos example might or might not help.
- code: https://github.com/pmndrs/jotai/tree/main/examples/todos
- playground: https://githubbox.com/pmndrs/jotai/tree/main/examples/todos

**Reply by lveillard** [reply]
It's almost there but with the difference that in the todos example we can store the current targeted todo's id in an atom.

in the values example we need to:
- Call multiple ids at once in different contexts
- Use values on other values (for instance todo-3's data is computed with values todo-2 which uses data from todo-1)

We are using familyAtoms now but will try to use a combination of this:
```
export const valueAtom = (fieldId: string) =>
  atom((get) => interpretExpression(get(fieldMapMap).get(fieldId).expression, dependencies.map((valueId)=>get(valueAtom(valueId)));
 ```
 and useMemo
 
 So basically a recursive atom that gets the expression at each level and gets or computes the nested dependencies as well
 
 Not sure if it will work like the family atom works now, but using the family atom is scary with all those memory warnings 😂

**Reply by dai-shi** [reply]
atomFamily isn't scary if it's used globally and you are sure that the params are not infinite.
But, in general, I think we could achieve the same behavior with atoms-in-atom.

**Reply by lveillard** [reply]
Hello @dai-shi and thank you again for all the brainpower you share with us :)

To refresh the contet, the use case 100% of the times is to send a prop of the react component directly as a parameter of the atom with parameter.

I managed to remove some atom families using the suggested pattern:
`export const valueAtom = (fieldId: string) =>
  atom((get) => .... `*
  
However this seems to do the same as the atomFamily and create new atoms each time a value is updated, which I'm not sure is normal.
In the image we can see how atom 82 migrated to atom 133 instead of just update atom 82. 

So each time I update the value, a bunch of atoms are created, so if i keep editing i get to >2000 atoms really fast. This does not seem to impact the performance, but I would like to know if is normal behaviour. 

Some of the react components load these atoms with parameters like this =
```tsx
//no new atoms here
My Comp = () => {
	const unitFields = useAtomValue(useMemo(() => unitAvailableFieldsAtom(unitId), [unitId]));
        ....
}
```

in other components we call them like this
```tsx
//this is calling atom()
	const currentField = useAtomValue(
		useMemo(() => atom((get) => get(dataFieldAtom(fieldId)) ... additional logic), [fieldId]),
	);
```

And sometimes we need loadable:

```tsx
export const MyComp = (unitid, fieldId) => {
//this is an atom with parameter with async magic happening inside
	const loadableValue = useAtomValue(useMemo(() => loadable(valueAtom(unitId, fieldId), [unitId, fieldId])
}
```

I would love to know if any of these are black patterns or if it is fine to create more atoms in each render.

The logic works perfectly even if new atoms are created.

  
  
![image](https://github.com/pmndrs/jotai/assets/29461076/d9b85ee0-3449-4613-a542-2195e7daaf78)

**Reply by dai-shi** [reply]
Creating a new atom is very lightweight, so if it's necessary, you don't need worry too much.
But, if it's possible to reduce the number of atoms, it may (or may not very much) improve the performance.

I would consider smaller atoms instead of a big atom, if it's reasonable.

**Reply by lveillard** [reply]
I see thanks. In this case we could do multiple little atoms but the number of atoms os unknown as the storage grows dynamically with interactions and database data.

By the way, is it normal to have those new atoms getting the data from the other atoms? For me ot looks weird to see that transfer of data from one atom group to the new atoms even if everything works as expected

**Reply by dai-shi** [reply]
I'm not 100% sure what you mean by the data transferring, but in a _pure_ Jotai atom model, we wouldn't need `useMemo(() => atom(...`. It is only needed to integrate with other React states. In such cases, it can't be helped.

**Reply by lveillard** [reply]
I meant __unlabeled-atom82__'s content migrating to the new atom __unlabeled-atom133__

I think it's the atm generators `const myAtom = (param) => atom( (get) => { .... stuff that uses the param)} }`

Anyway as long as i'm not doing any black pattern that might not work in the long term or degrade the performance, i'm happy! Thanks again for the help
